Add option for clients to seek head blob first when downloading

kaykurokawa opened the pull request:
This adds an option in conf for clients to download by prioritizing the seeking of the "head blob" (the very first blob in the stream) before they seek the next unavailable blob. 

It is turned off by default in this PR so will not change how downloading works unless configured to be turned on manually.  Should be turned on when reflector switch to announcing head blob only.

tzarebczan commented:
In what cases should this option be turned on? How would it be turned on manually?

kaykurokawa commented:
"seek head blob first" download method should not be turned on until our reflector servers has switched to announcing head blob only.

It can be turned on and off through the configuration file (the default configuration is set to off)

jackrobison commented on lbrynet/core/client/ConnectionManager.py:
```diff
@@ -146,57 +150,58 @@ def _send_secondary_requests(self, peer, protocol):
+            return []
+        out = [peer for peer in peers if peer not in self._peer_connections]
+        random.shuffle(out)
+        out = out[0:new_conns_needed]
```
This can just be `return out[:new_conns_needed]`

jackrobison commented on lbrynet/core/client/ConnectionManager.py:
```diff
@@ -146,57 +150,58 @@ def _send_secondary_requests(self, peer, protocol):
+    def return_shuffled_peers_not_connected_to(self, peers, new_conns_needed):
+        if peers is None:
+            # can happen if there is some error in the lookup
+            return []
```
Should this be an exception? A warning?

jackrobison commented on lbrynet/core/client/ConnectionManager.py:
```diff
@@ -146,57 +150,58 @@ def _send_secondary_requests(self, peer, protocol):
-        # TODO: just pick the first to which we don't have an open connection
+
+        # find peers for the head blob if configured to do so
+        if self.seek_head_blob_first is True:
```
`if self.seek_head_blob_first:`

jackrobison commented on lbrynet/core/client/ConnectionManager.py:
```diff
@@ -146,57 +150,58 @@ def _send_secondary_requests(self, peer, protocol):
-    def _get_new_peers(self, request_creators):
+    def _get_new_peers(self):
+        new_conns_needed = self.max_connections_per_stream - len(self._peer_connections)
+        if new_conns_needed < 1:
```
`if not new_conns_needed:`

lyoshenka replied:
@jackrobison what if there's a bug and `new_conns_needed` ends up being -1 somehow? :smiling_imp:

jackrobison commented on lbrynet/core/client/ConnectionManager.py:
```diff
@@ -146,57 +150,58 @@ def _send_secondary_requests(self, peer, protocol):
+
+        # we didn't find any new peers on the head blob,
+        # we have to look for the first unavailable blob
+        if len(peers) == 0:
```
`if not peers:`

jackrobison commented on lbrynet/core/client/ConnectionManager.py:
```diff
@@ -146,57 +150,58 @@ def _send_secondary_requests(self, peer, protocol):
-
-        return sorted(self._primary_request_creators, key=count_peers)
+    def return_shuffled_peers_not_connected_to(self, peers, new_conns_needed):
+        if peers is None:
```
Should this be `if not peers:`? When is `peers` None?

kaykurokawa replied:
this right here: 
https://github.com/lbryio/lbry/blob/master/lbrynet/core/client/BlobRequester.py#L118
It looks like you could get that when there is no more blobs for it to download... not sure.  I'll look through the code again to see if we can improve this part.

kaykurokawa commented:
rebased to master
made pep8 fixes as requested by jack 
Will need to check  BlobRequest._find_peers_for_hash() returning None

kaykurokawa commented:
Corrected and clarified in docstrings the returning of None when looking for peers. 

BlobRequestr._get_hash_for_peer_search() , function to get the next unavailable hash, could return None if there was nothing left to download. Previously, the None would propagate and functions would have its own way of handling None. Now we deal with None immediately in BlobRequester. get_new_peers_for_next_unavailable() and removed all other handling of None in other places.

lyoshenka commented:
@kaykurokawa whats the next step on this?

kaykurokawa commented:
@jackrobison please take a look at 734c5a1
